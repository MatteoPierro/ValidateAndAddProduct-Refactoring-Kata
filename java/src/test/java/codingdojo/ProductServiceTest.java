package codingdojo;

import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import static org.approvaltests.Approvals.verify;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class ProductServiceTest {
    @Test
    void validateAndAdd() {

        CombinationApprovals.verifyAllCombinations(this::doValidateAndAdd,
                new String[]{"", "Sample product", "Sample Queen product"},
                new String[]{"", "Lipstick", "Eyeshadow", "Mascara", "Blusher", "Foundation", "Unknown"},
                new Double[]{-1D, 5D, 11D},
                new Double[]{10D, 11D, 21D, 26D},
                new Boolean[]{false, true}
                );
    }

    private String doValidateAndAdd(String name, String type, Double weight, Double price, Boolean recyclable) {
        ProductFormData productData = new ProductFormData(name, type, weight, price, recyclable);
        FakeDatabase db = new FakeDatabase();
        ProductService sut = new ProductService(db);
        Response response = sut.validateAndAdd(productData);

        // Assert
        String productString = "";
        if (db.product != null) {
            productString = db.product.toString();
        }
        return response.toString() + " " + productString;
    }

    class FakeDatabase extends DatabaseAccess {
        private Product product;

        @Override
        public int storeProduct(Product product) {
            this.product = product;
            return 1;
        }
    }

}
