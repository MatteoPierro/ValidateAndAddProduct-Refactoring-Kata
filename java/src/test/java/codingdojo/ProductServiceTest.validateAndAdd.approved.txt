Response{productId=1, statusCode=0, message='Product Successfully Added'} Product{name='Sample product', type='Lipstick', weight=5.0, family=LIPS, range=BUDGET}